### PR TITLE
Skip to create and bind ClusterRole if already there

### PR DESCRIPTION
**Pull request by wangkexiong**
for Multiple Aerospike clusters in multiple k8s namespaces

### CODE REVIEW

**Comment by spkesan** [comment]
Thanks @wangkexiong for the PR.
Did you test it out ?

The `ServiceAccount` created in a new namespace must also be added to the existing `ClusterRoleBinding`. Otherwise, the new service account in the new namespace will not have any roles bound to it.
Also, I don't think it's possible to modify an existing resource via helm. So, we must create a new `ClusterRoleBinding` with a unique name for each new service account.

**Comment by spkesan** [comment]
I think we also cannot use a `RoleBinding` instead of `ClusterRoleBinding` because the init container queries the list of Kubernetes nodes at cluster scope.

File: helm-charts/aerospike-cluster/templates/aerospike-cluster-rbac.yaml [hunk 1]
```diff
@@ -11,6 +11,7 @@ metadata:
     release: {{ .Release.Name }}
 
 ---
+{{- if not (lookup "rbac.authorization.k8s.io/v1" "ClusterRole" "" "aerospike-cluster") }}
```
Comment (spkesan) [inline helm-charts/aerospike-cluster/templates/aerospike-cluster-rbac.yaml:14]: Can you make this condition just for `ClusterRole`?
Like this,

```
{{- if not (lookup "rbac.authorization.k8s.io/v1" "ClusterRole" "" "aerospike-cluster") }}
# Role
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: aerospike-cluster
  labels:
    app: {{ template "aerospike-cluster.commonName" . }}
    chart: {{ .Chart.Name }}
    release: {{ .Release.Name }}
rules:
- apiGroups:
  - ""
  resources:
  - pods
  - nodes
  - services
  verbs:
  - get
  - list
- apiGroups:
  - aerospike.com
  resources:
  - '*'
  verbs:
  - '*'
{{- end }}
```

And also change `ClusterRoleBinding` name to `aerospike-cluster-{{ .Release.Namespace }}`.
Like this,
```
# RoleBinding
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: aerospike-cluster-{{ .Release.Namespace }}
  labels:
    app: {{ template "aerospike-cluster.commonName" . }}
    chart: {{ .Chart.Name }}
    release: {{ .Release.Name }}
roleRef:
  kind: ClusterRole
  name: aerospike-cluster
  apiGroup: rbac.authorization.k8s.io
subjects:
- kind: ServiceAccount
  name: aerospike-cluster
  namespace: {{ .Release.Namespace }}
```

**Review by spkesan — CHANGES_REQUESTED** [review]

**Comment by wangkexiong** [comment]
I tested with the change to skip ClusterRole and ClusterRoleBinding, it really works. But it is NOT a good solution...
The subject section in ClusterRoleBinding only includes the first namespace (We skipped to create it again if it is already there)
I did not find any way to update this section to include the new namespace.

Test bed: Docker Desktop 2.5.0.1 on Win10, K8S version: v1.19.3
I packed the changes on my personal repo

```bash
$ kubectl create ns aerospike
$ helm install aerospike-kubernetes-operator personal/aerospike-kubernetes-operator -f operator.yaml -n aerospike
$ kubectl create ns aerospike1
$ kubectl create ns test
$ kubectl create secret generic aerospike-secrets --from-file=features.conf -n aerospike1
$ kubectl create secret generic auth-secret --from-literal=password=admin123 -n aerospike1
$ kubectl create secret generic auth-data --from-literal=password=data123 -n aerospike1
$ helm install aerocluster personal/aerospike-cluster -n aerospike1 -f cluster-singlenode.yaml
$ kubectl create secret generic aerospike-secrets --from-file=features.conf -n test
$ helm install aerocluster personal/aerospike-cluster -n test --set devMode=true,replicas=2,aerospikeSecretName=aerospike-secrets
$ kubectl get ClusterRoleBinding aerospike-cluster -o yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  annotations:
    meta.helm.sh/release-name: aerocluster
    meta.helm.sh/release-namespace: aerospike1
  creationTimestamp: "2021-05-06T06:33:38Z"
  labels:
    app: aerocluster
    app.kubernetes.io/managed-by: Helm
    chart: aerospike-cluster
    release: aerocluster
  managedFields:
  - apiVersion: rbac.authorization.k8s.io/v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:meta.helm.sh/release-name: {}
          f:meta.helm.sh/release-namespace: {}
        f:labels:
          .: {}
          f:app: {}
          f:app.kubernetes.io/managed-by: {}
          f:chart: {}
          f:release: {}
      f:roleRef:
        f:apiGroup: {}
        f:kind: {}
        f:name: {}
      f:subjects: {}
    manager: Go-http-client
    operation: Update
    time: "2021-05-06T06:33:38Z"
  name: aerospike-cluster
  resourceVersion: "2530"
  selfLink: /apis/rbac.authorization.k8s.io/v1/clusterrolebindings/aerospike-cluster
  uid: b14f28f6-c6b2-4644-b97e-625eb9316a07
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: aerospike-cluster
subjects:
- kind: ServiceAccount
  name: aerospike-cluster
  namespace: aerospike1
```

**Comment by wangkexiong** [comment]
Then I go to read the doc and think it again. Since we want to support multiple aerospike cluster in multiple namespaces.
We'd better NOT to use Cluster wide role and binding settings, and I change the code to use role and rolebinding for different namespace.

**Comment by spkesan** [comment]
@wangkexiong Apologies for the late reply.
Have you tested with just `Role` and `RoleBinding`? Does it work?
I believe we need a cluster wide access because Aerospike init container queries the list of kubernetes nodes at cluster scope.

**Comment by spkesan** [comment]
I think this will require some internal discussion and work to better address this issue. Putting this on hold for now.

**Comment by sud82** [comment]
As @spkesan said, we need cluster-wide access because Aerospike init container queries the list of Kubernetes nodes at a cluster scope. We will try to address the issue of the already existing ClusterRole in new helm charts based on new sdk

**Comment by sud82** [comment]
Resolved in new helm charts (AKO2.0)

**Pull request closed without merging**